crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -188,40 +214,354 @@ pub struct StateTransitionEvent<S: States> {
 
 /// Run the enter schedule (if it exists) for the current state.
 pub fn run_enter_schedule<S: States>(world: &mut World) {
-    world
-        .try_run_schedule(OnEnter(world.resource::<State<S>>().0.clone()))
-        .ok();
+    let Some(state) = world.get_resource::<State<S>>() else {
+        return;
+    };
+    let state = state.0.clone();
+    world.try_run_schedule(DeriveStates::<S>::default()).ok();
+    world.try_run_schedule(OnEnter(state)).ok();
 }
 
 /// If a new state is queued in [`NextState<S>`], this system:
 /// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
 /// - Sends a relevant [`StateTransitionEvent`]
 /// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
 /// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
+/// - Derive any derived states through the [`DeriveStates::<S>`] schedule, if it exists.
```

Comment:
This line will be much clearer if you don't use the same word "derive" repeatedly. Varying your language with synonyms is really helpful to readers who don't know / understand the exact word you're using.